[DataPipe] only apply special serialization when dill is installed

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74958

This is a quick fix. Only applies the custom serialization logic when `dill` is installed. The specific issue mentioned below should be resolved if `dill` is not installed in the local environment.

See the [issue in this comment](https://github.com/pytorch/data/issues/237#issuecomment-1080651807).

I also plan to add some test related to circular dependency in the serialization process.

NivekT committed:
[DataPipe] only apply special serialization when dill is installed

[ghstack-poisoned]

ejguan reviewed (commented):
Could we add a test for circular dependency to prevent future changes violating this behavior

NivekT committed:
Update on "[DataPipe] only apply special serialization when dill is installed"


This is a quick fix. Only applies the custom serialization logic when `dill` is installed. The specific issue mentioned below should be resolved if `dill` is not installed in the local environment.

See the [issue in this comment](https://github.com/pytorch/data/issues/237#issuecomment-1080651807).

I also plan to add some test related to circular dependency in the serialization process.

[ghstack-poisoned]

NivekT commented:
> Could we add a test for circular dependency to prevent future changes violating this behavior

Just added the case described in the issue. I am still looking into `dill` related problem.

NivekT commented:
@pytorchbot merge this please

pmeier commented:
In addition, datapipes that spawn `_ChildDataPipe`'s also have a custom serialization logic that needs to be fixed:

- https://github.com/pytorch/pytorch/blob/79307fbde0733aef7a8d73f3cebfabc60a70110d/torch/utils/data/datapipes/iter/combining.py#L158-L167
- https://github.com/pytorch/pytorch/blob/79307fbde0733aef7a8d73f3cebfabc60a70110d/torch/utils/data/datapipes/iter/combining.py#L343-L355

Otherwise, we still get an infinite recursion:

```py
import pickle

from torchdata.datapipes.iter import IterDataPipe, IterableWrapper, Zipper


class CustomIterDataPipe(IterDataPipe):
    def classify(self, x):
        return 0

    def __init__(self):
        self._dp = Zipper(*IterableWrapper([]).demux(2, self.classify))

    def __iter__(self):
        yield from self._dp


pickle.dumps(CustomIterDataPipe())
```

ejguan commented:
> Could you look into why traverse is not failing on local functions while pickle.dumps does?

This is due to we have `only_datapipe=True` argument in `traverse` function to remove all the bounded functions from pickling. But, it raises another issue with `traverse`. Since we now can rely on pickle to determine the circular reference problem, `traverse(datapipe, only_datapipe=False)` won't handle the circular reference issue.

ejguan commented:
> > Could you look into why traverse is not failing on local functions while pickle.dumps does?
> 
> This is due to we have `only_datapipe=True` argument in `traverse` function to remove all the bounded functions from pickling. But, it raises another issue with `traverse`. Since we now can rely on pickle to determine the circular reference problem, `traverse(datapipe, only_datapipe=False)` won't handle the circular reference issue.

I find the way to fix it. Sending a PR.

pmeier commented:
> This is due to we have `only_datapipe=True` argument in `traverse` function to remove all the bounded functions from pickling.

But doesn't that mean, that we can use `lambda`'s and local functions since they are not touched by the `DataLoader2` anyway? 

As explained in https://github.com/pytorch/pytorch/pull/74958#issuecomment-1084338879 the `torchvision` test started with `traverse`, but that didn't detect local functions. In our sync someone from the torchdata team (I don't recall) told us that local functions are forbidden without `dill` (is this still true? If yes, why doesn't traverse fail?) and so we switched the test to `pickle.dumps`, which fails as we wanted it to do. But this discussion seems that this again is not what we actually should be testing. 

We need to test if our datasets are fully compatible with real world uses together with the `DataLoader2`. How would such a test look like?

```py
dataset = ...

dp = dataset.batch(2).collate()

# Maybe we can make this is a static method of the DataLoader2?
try:
    num_workers = len(os.sched_getaffinity(0))
except Exception:
    num_workers = os.cpu_count() or 1

for parallelism_mode in ["mp", "thread"]:
    dl = DataLoader2(
        dp,
        batch_size=None,
        shuffle=True,
        num_workers=num_workers,
        parallelism_mode=parallelism_mode,
        timeout=1,
    )

    for _ in dl:
        pass
```

ejguan commented:
Looked into the codebase. I think the bug actually comes from this line https://github.com/pytorch/pytorch/blob/00607e75e5f3157839719efb5b26ca92e91190a3/torch/utils/data/_utils/serialization.py#L29

When we pickle a DataPipe, it will pickle the function with new `cache` IMHO. Then, when pickle the function, the original DataPipe will be `pickled` with the new `cache` so this causes infinite recursion.

If we change the above line to `return fn, SerializationType("pickle")`, this should be resolved.
